src/analysis/mean_inequalities.lean

```diff
@@ -0,0 +1,147 @@
+/-
+Copyright (c) 2019 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import analysis.calculus.mean_value data.nat.parity analysis.complex.exponential
+  analysis.convex.specific_functions
+
+/-!
+# Mean values inequalities
+
+In this file we prove various inequalities between mean values:
+arithmetic mean, geometric mean, generalized mean (natural and integer
+cases).
+
+For generalized means we only prove
+$\left(∑_jw_jz_j\right)^n≤∑_jw_jz_j^n$ because standard versions would
+require $\sqrt[n]{x}$ which is not implemented in `mathlib` yet.
+
+Probably a better approach to the generalized means inequality is to
+prove `convex_on_rpow` in `analysis/convex/specific_functions` first,
+then apply it.
+
+It is not yet clear which versions will be useful in the future, so we
+provide a few different two forms of most inequalities : `ℝ` vs
+`ℝ≥0`. For the AM-GM inequality we also prove special cases for `n=2`
+and `n=3`.
+-/
+
+universes u v
+
+open real finset
+open_locale classical nnreal
+
+variables {ι : Type u} (s : finset ι)
+
+/-- Geometric mean is less than or equal to the arithmetic mean, weighted version
+for functions on `finset`s. -/
+theorem real.am_gm_weighted (w z : ι → ℝ)
+  (hw : ∀ i ∈ s, 0 ≤ w i) (hw' : s.sum w = 1) (hz : ∀ i ∈ s, 0 ≤ z i) :
+  s.prod (λ i, (z i) ^ (w i)) ≤ s.sum (λ i, w i * z i) :=
+begin
+  let s' := s.filter (λ i, w i ≠ 0),
+  rw [← sum_filter_ne_zero] at hw',
+  suffices : s'.prod (λ i, (z i) ^ (w i)) ≤ s'.sum (λ i, w i * z i),
+  { have A : ∀ i ∈ s, i ∉ s' → w i = 0,
+    { intros i hi hi',
+      simpa only [hi, mem_filter, ne.def, true_and, not_not] using hi' },
+    have B : ∀ i ∈ s, i ∉ s' → (z i) ^ (w i) = 1,
+      from λ i hi hi', by rw [A i hi hi', rpow_zero],
+    have C : ∀ i ∈ s, i ∉ s' → w i * z i = 0,
+      from λ i hi hi', by rw [A i hi hi', zero_mul],
+    rwa [← prod_subset s.filter_subset B, ← sum_subset s.filter_subset C] },
+  replace A : ∀ i ∈ s', i ∈ s ∧ w i ≠ 0, from λ i hi, mem_filter.1 hi,
+  replace hz : ∀ i ∈ s', 0 ≤ z i := λ i hi, hz i (A i hi).1,
+  replace hw : ∀ i ∈ s', 0 ≤ w i := λ i hi, hw i (A i hi).1,
+  by_cases B : ∃ i ∈ s', z i = 0,
+  { rcases B with ⟨i, imem, hzi⟩,
+    rw [prod_eq_zero imem],
+    { exact sum_nonneg (λ j hj, mul_nonneg (hw j hj) (hz j hj)) },
+    { rw hzi, exact zero_rpow (A i imem).2 } },
+  { replace hz : ∀ i ∈ s', 0 < z i,
+      from λ i hi, lt_of_le_of_ne (hz _ hi) (λ h, B ⟨i, hi, h.symm⟩),
+    have := convex_on_exp.map_sum_le hw hw' (λ i _, set.mem_univ $ log (z i)),
+    simp only [exp_sum, (∘), smul_eq_mul, mul_comm (w _) (log _)] at this,
+    convert this using 1,
+    { exact prod_congr rfl (λ i hi, rpow_def_of_pos (hz i hi) _) },
+    { exact sum_congr rfl (λ i hi, congr_arg _ (exp_log $ hz i hi).symm) } }
+end
+
+theorem nnreal.am_gm_weighted (w z : ι → ℝ≥0) (hw' : s.sum w = 1) :
+  s.prod (λ i, (z i:ℝ) ^ (w i : ℝ)) ≤ s.sum (λ i, w i * z i) :=
```

Comment:
the spacing around colons in coercions is not coherent.

Reply:
Fixed.